ga4gh/datamodel/references.py

```diff
@@ -361,75 +399,80 @@ class HtslibReferenceSet(datamodel.PysamDatamodelMixin, AbstractReferenceSet):
     """
     A referenceSet based on data on a file system
     """
-    def __init__(self, localId, dataDir, dataRepository):
+    def __init__(self, localId):
         super(HtslibReferenceSet, self).__init__(localId)
-        self._dataDir = dataDir
-        self._setMetadata()
-        self._scanDataFiles(dataDir, ["*.fa.gz"])
-
-    def _setMetadata(self):
-        metadataFileName = '{}.json'.format(self._dataDir)
-        with open(metadataFileName) as metadataFile:
-            metadata = json.load(metadataFile)
-            try:
-                self._assemblyId = metadata['assemblyId']
-                self._description = metadata['description']
-                self._isDerived = metadata['isDerived']
-                self._ncbiTaxonId = metadata['ncbiTaxonId']
-                self._sourceAccessions = metadata['sourceAccessions']
-                self._sourceUri = metadata['sourceUri']
-            except KeyError as err:
-                raise exceptions.MissingReferenceSetMetadata(
-                    metadataFileName, str(err))
-
-    def _addDataFile(self, path):
-        dirname, filename = os.path.split(path)
-        localId = ".".join(filename.split(".")[:-2])
-        metadataFileName = os.path.join(dirname, "{}.json".format(localId))
-        with open(metadataFileName) as metadataFile:
-            metadata = json.load(metadataFile)
-        reference = HtslibReference(self, localId, path, metadata)
-        self.addReference(reference)
+        self._dataUrl = None
+
+    def populateFromFile(self, dataUrl):
+        """
+        Populates the instance variables of this ReferencSet from the
+        data URL.
+        """
+        self._dataUrl = dataUrl
+        fastaFile = self.getFastaFile()
+        for referenceName in fastaFile.references:
+            reference = HtslibReference(self, referenceName)
+            # TODO break this up into chunks and calculate the MD5
+            # in bits (say, 64K chunks?)
+            bases = fastaFile.fetch(referenceName)
+            md5checksum = hashlib.md5(bases).hexdigest()
+            reference.setMd5checksum(md5checksum)
+            reference.setLength(len(bases))
+            self.addReference(reference)
+
+    def populateFromRow(self, row):
+        """
+        Populates this reference set from the values in the specified DB
+        row.
+        """
+        self._dataUrl = row[b'dataUrl']
+        self._description = row[b'description']
+        self._assemblyId = row[b'assemblyId']
+        self._isDerived = bool(row[b'isDerived'])
+        self._md5checksum = row[b'md5checksum']
+        self._ncbiTaxonId = row[b'ncbiTaxonId']
+        self._sourceAccessions = json.loads(row[b'sourceAccessions'])
```

Comment:
I'm not sure sourceAccessions are ever set for single references. https://github.com/ga4gh/server/issues/1185